packages/arb-evm/message/data.go

```diff
@@ -67,3 +72,196 @@ func marshaledBytesHash(data []byte) common.Hash {
 	}
 	return ret
 }
+
+func encodeAmount(amount *big.Int) ([]byte, error) {
+	mod := big.NewInt(10)
+	zero := big.NewInt(0)
+	exp := byte(0)
+	for amount.Cmp(zero) > 0 && new(big.Int).Mod(amount, mod).Cmp(zero) == 0 {
+		amount = amount.Div(amount, mod)
+		exp++
+	}
+	amountData, err := rlp.EncodeToBytes(amount)
+	if err != nil {
+		return nil, err
+	}
+
+	if amount.Cmp(zero) == 0 {
+		return amountData, nil
+	}
+	return append(amountData, exp), nil
+}
+
+func decodeAmount(r io.Reader) (*big.Int, error) {
+	a := new(big.Int)
+	if err := rlp.Decode(r, a); err != nil {
+		return nil, err
+	}
+	if a.Cmp(big.NewInt(0)) == 0 {
+		return a, nil
+	}
+	bData := make([]byte, 1)
+	n, _ := r.Read(bData)
+	if n != 1 {
+		return nil, errors.New("not enough data for exponent in amount")
+	}
+	amount := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(bData[0])), nil)
+	return amount.Mul(amount, a), nil
+}
+
+type CompressedAddress interface {
+	fmt.Stringer
+	isCompressedAddress()
+	Encode() ([]byte, error)
+}
+
+type CompressedAddressIndex struct {
+	*big.Int
+}
+
+func (c CompressedAddressIndex) isCompressedAddress() {}
+
+func (c CompressedAddressIndex) Encode() ([]byte, error) {
+	return rlp.EncodeToBytes(c.Int)
+}
+
+func (c CompressedAddressIndex) String() string {
+	return fmt.Sprintf("Index[%v]", c)
+}
+
+type CompressedAddressFull struct {
+	common.Address
+}
+
+func (c CompressedAddressFull) isCompressedAddress() {}
+
+func (c CompressedAddressFull) Encode() ([]byte, error) {
+	return rlp.EncodeToBytes(c.Bytes())
+}
+
+func (c CompressedAddressFull) String() string {
+	return fmt.Sprintf("Address[%v]", c.Hex())
+}
+
+func decodeAddress(r io.Reader) (CompressedAddress, error) {
```

Comment:
I don't see how this function can work. Won't the first rlp.Decode always succeed?

Reply:
I'm fairly certain that it'll only succeed if the RLP value is an encoded integer rather than an encoded byte slice, but I'll add some tests to confirm `CompressedAddress` works as intended

Reply:
And I was totally wrong about that 😁 

I've now added tests for all 3 cases and everything is passing perfectly